crypto/src/mls/credential/trust_anchor.rs

```diff
@@ -0,0 +1,1647 @@
+        let certificate_chain = self.validate(backend, group_context)?;
+        Ok(openmls::extensions::PerDomainTrustAnchor::new(
+            self.domain_name.into(),
+            openmls::prelude::CredentialType::X509,
+            certificate_chain,
+        )
+        .map_err(MlsError::from)?)
+    }
+}
+
+impl PerDomainTrustAnchor {
+    /// Validates the trust anchor and return its encoded chain encoded to der.
+    fn validate(
+        &self,
+        backend: &MlsCryptoProvider,
+        group_context: Option<&GroupContext>,
+    ) -> CryptoResult<Vec<Vec<u8>>> {
+        // parse PEM
+        let certificate_chain = pem::parse_many(&self.intermediate_certificate_chain)?
+            .iter()
+            .map(|p| Certificate::from_der(p.contents()))
+            .collect::<Result<PkiPath, x509_cert::der::Error>>()?;
+
+        // verify domain_name is unique
+        if let Some(group_context) = group_context {
+            if group_context
+                .extensions()
+                .per_domain_trust_anchors()
+                .is_some_and(|anchors| {
+                    anchors
+                        .iter()
+                        .any(|anchor| String::from_utf8_lossy(anchor.domain_name()) == self.domain_name)
+                })
+            {
+                return Err(CryptoError::DuplicateDomainName);
+            }
+        }
+
+        // self signed or incomplete chains are not allowed
+        let leaf_cert = certificate_chain.first().ok_or(CryptoError::InvalidCertificateChain)?;
+
+        // validate domain in the leaf matches with the one supplied
+        let domain_name = extract_domain_name(leaf_cert)?;
+        if domain_name != self.domain_name {
+            return Err(CryptoError::DomainNamesDontMatch);
+        }
+
+        // verify the whole chain
+        let root_cert = certificate_chain
+            .iter()
+            .map(Ok)
+            .reduce(
+                |child, parent| -> Result<&Certificate, openmls_traits::types::CryptoError> {
+                    let child = child?;
+                    let parent = parent?;
+                    child.is_valid()?;
+                    child.is_signed_by(backend.crypto(), parent)?;
+                    Ok(parent)
+                },
+            )
+            .unwrap()
+            .map_err(MlsError::from)?;
+        // ensure that the root is also valid
+        root_cert.is_valid().map_err(MlsError::from)?;
+
+        check_root_in_trust_store(root_cert)?;
+
+        let encoded_chain = certificate_chain
+            .into_iter()
+            .map(|cert| cert.to_der())
+            .collect::<Result<_, x509_cert::der::Error>>()?;
+        Ok(encoded_chain)
+    }
+}
+
+/// Checks the root cert against the trust store. In wasm maybe use webpki-roots (https://github.com/rustls/webpki-roots) crate
+fn check_root_in_trust_store(_root: &Certificate) -> CryptoResult<()> {
+    // TODO: verify root certificate is valid in the device's trust store
```

Comment:
Is this verification supposed to be implemented or is this out of scope of this PR?

Reply:
It is out of the scope of this PR.